### PR TITLE
Feature/69 webscraper

### PR DESCRIPTION
**Pull request by Yfkesmit**
Hoera! Een stabiele versie van webscraper

pls kijk naar alle files van webscraper en de tests en misschien ook wel de filter koppeling en check op schoonheidsfoutjes.

Development is nog niet in webscraper gemerged voor het overzicht, wanneer changes zijn approved moet dus eerst dev in web worden gemerged voordat web in dev kan.

### CODE REVIEW

File: scraping/web_scraper/README.md [hunk 1]
```diff
@@ -1,34 +1,31 @@
...
+The script for the EC website has two different functionalities. First, it fetches a list of medicines to scrape. For each of these medicines, a folder is created where the necessary information about the medicine is temporarily stored. In addition, the EC scraper actually ensures that all necessary information is retrieved from the EC website. This information consists of a number of data points for a medicine (attributes), links to additional procedure files (decisions and annexes) and links to the EMA website. The attributes of a medicines are stored in the aforementioned folder, in a JSON. The links to decision documents and annex documents for all medicines are stored in a single JSON (urls.json). This file also contains links to the EMA and the EMA scraper will add to this file too.
```
Comment (elioverhoef) [inline scraping/web_scraper/README.md:9]: the attributes of a medicines -> medicine

File: scraping/web_scraper/filter_retry.py [hunk 2]
```diff
@@ -52,22 +58,23 @@ def retry_all(filter_path: str, urls_file: dict[str, dict[str, list[str]]], data
             if eu_n in urls_file:
                 m.get_urls_ec(urls_file[eu_n]["ec_url"], eu_n, med_type_dict[''.join(filename_elements[:2])],
```
Comment (elioverhoef) [inline scraping/web_scraper/filter_retry.py:59]: get_urls_ec wil één url als string, krijgt vanuit urls_file[eu_n]["ec_url"] lijst van strings. Misschien [0] toevoegen?

File: scraping/web_scraper/filter_retry.py [hunk 3]
```diff
@@ -27,10 +27,16 @@ def run_filter(n: int, data_filepath: str = "../../data"):
         n (int): Number of times filter is called
         data_filepath (str, optional): the path to the data folder. Defaults to "../../data"
     """
+    json_path = "web_scraper/"
+    filter_path = ""
+    # If file is run locally:
+    if "web_scraper" in os.getcwd():
+        json_path = ""
+        filter_path = "../"
     for i in range(n):
```
Comment (elioverhoef) [inline scraping/web_scraper/filter_retry.py:36]: i not used. Replace by _?

**Review by elioverhoef — APPROVED** [review]
Looking good! Alleen nog PEP8/typing dingetjes

File: scraping/filter/pdf_filter.py [hunk 4]
```diff
@@ -175,7 +175,11 @@ def get_url(filename) -> str:
     """
     eu_num = filename.split('_')[0]
     try:
-        with open('../web_scraper/JSON/urls.json') as urls_json:
+        json_path = "web_scraper/"
+        # If file is run from webscraper locally:
+        if "web_scraper" in os.getcwd():
```
Comment (matthboers) [inline scraping/filter/pdf_filter.py:180]: Could be more explicit by specifying that the user is running from the `web_scraper/` directory?
(Change `in` to something that checks that the last dir == web_scraper)

File: scraping/web_scraper/filter_retry.py [hunk 5]
```diff
@@ -27,10 +27,16 @@ def run_filter(n: int, data_filepath: str = "../../data"):
         n (int): Number of times filter is called
         data_filepath (str, optional): the path to the data folder. Defaults to "../../data"
     """
+    json_path = "web_scraper/"
+    filter_path = ""
+    # If file is run locally:
+    if "web_scraper" in os.getcwd():
```
Comment (matthboers) [inline scraping/web_scraper/filter_retry.py:33]: Same potential issue as https://github.com/MediSeeUU/MediSeeUU/pull/16#discussion_r1020184583

**Review by matthboers — APPROVED** [review]
Found issues are minor, but great if they could be fixed before merging. Everything looks good otherwise!

**Review by miloravi — APPROVED** [review]

File: scraping/web_scraper/filter_retry.py (hunk 5 above)
Comment (matthboers) [reply scraping/web_scraper/filter_retry.py]: Solved in e95c26d08a250ca7a620b437be98d0ae16087eb4

File: scraping/filter/pdf_filter.py (hunk 4 above)
Comment (matthboers) [reply scraping/filter/pdf_filter.py]: Solved in e95c26d08a250ca7a620b437be98d0ae16087eb4

File: scraping/web_scraper/filter_retry.py (hunk 3 above)
Comment (matthboers) [reply scraping/web_scraper/filter_retry.py]: Fixed in 232b78e07b87e90ba56cd46794d0c02393bfe576

File: scraping/web_scraper/README.md (hunk 1 above)
Comment (matthboers) [reply scraping/web_scraper/README.md]: Fixed in 232b78e07b87e90ba56cd46794d0c02393bfe576

File: scraping/web_scraper/filter_retry.py (hunk 2 above)
Comment (matthboers) [reply scraping/web_scraper/filter_retry.py:61]: Fixed in 232b78e07b87e90ba56cd46794d0c02393bfe576

File: scraping/web_scraper/__main__.py [hunk 6]
```diff
@@ -190,11 +187,11 @@ def main(data_filepath: str = '../data'):
 
         with tqdm_logging.logging_redirect_tqdm():
             if use_parallelization:
-                tqdm_concurrent.thread_map(get_urls_ema_retry, *unzipped_ema_urls)
+                tqdm_concurrent.thread_map(get_urls_ema, *unzipped_ema_urls, max_workers=12)
```
Comment (XiaoyiHu1998) [inline scraping/web_scraper/__main__.py:190]: je kunt max workers dynamisch aan corecount aanpassen met een import, refer to pdf_scraper main for example

File: scraping/web_scraper/utils.py [hunk 7]
```diff
@@ -5,39 +5,52 @@
...
     """
-    def wrapper(*args, **kwargs):
-        exception_names: list[str] = []
-        if func is None:
-            return None
+    # Decorators are a little tricky. if you want to pass arguments to a function with the @exception_retry notation,
+    # You need to split the decorator into two functions. The first function catches the arguments, and returns a
+    # function. The second calls the first function with the argument of the function that has to be retried.
+    # A bit messy, but it is what it is with Python.
+    def decorator(func: callable):
```
Comment (XiaoyiHu1998) [inline scraping/web_scraper/utils.py:23]: is this a general decorator or for a specific usecase? if for specific case maybe consider making a function that takes different paramters and transforms them to what you would throw into the original function or modify the return result of the original

Comment (XiaoyiHu1998) [reply scraping/web_scraper/utils.py:23]: dont need to change during this pull request tho

**Review by XiaoyiHu1998 — APPROVED** [review]

**Pull request merged**